mlir/lib/Dialect/Tosa/Transforms/TosaFolders.cpp

```diff
@@ -1013,6 +1013,68 @@ struct TosaFoldConstantGreater : public TosaFoldConstantBinary<TosaFoldConstantG
   }
 };
 
+struct TosaFoldConstantBitwiseNot
+    : public TosaFoldConstantUnaryElementwise<TosaFoldConstantBitwiseNot,
+                                              BitwiseNotOp> {
+  using TosaFoldConstantUnaryElementwise<
+      TosaFoldConstantBitwiseNot,
+      BitwiseNotOp>::TosaFoldConstantUnaryElementwise;
+
+  DenseElementsAttr computeInteger(DenseElementsAttr values,
+                                   PatternRewriter &rewriter, TosaOp op) const {
+    return applyElementWise<APInt, APInt, IntegerType>(
+        values,
+        [](const APInt &val, IntegerType) {
+          return APInt(val.getBitWidth(), ~val.getZExtValue());
+        },
+        cast<IntegerType>(values.getElementType()));
+  }
+};
+
+struct TosaFoldConstantCeil
+    : public TosaFoldConstantUnaryElementwise<TosaFoldConstantCeil, CeilOp> {
+  using TosaFoldConstantUnaryElementwise<
+      TosaFoldConstantCeil, CeilOp>::TosaFoldConstantUnaryElementwise;
+
+  DenseElementsAttr computeFloat(DenseElementsAttr values,
+                                 PatternRewriter &rewriter, TosaOp op) const {
+    return applyElementWise<APFloat, APFloat, FloatType>(
+        values,
+        [](const APFloat &val, FloatType) {
+          // Compute ceil (APFloat unfortunately does not provide this function,
+          // such that we need to unpack here)
+          auto res = APFloat(std::ceil(val.convertToFloat()));
```

Comment:
Let's try avoiding converting to float (which would be wrong if the input is double).
```suggestion
          APFloat ret = val;
          val.roundToIntegral(RoundingMode::TowardPositive);
          return val;
```